### PR TITLE
feat: get collision between bodies

### PR DESCRIPTION
**Pull request by umutsoysalansys**
Features added:
 - distance between two objects. 
 - Collision State between two objects (None, Touch, Intersect, confined)

### CODE REVIEW

**Comment by RobPasMue** [comment]
Hi @umutsoysalansys - new API package is available. I updated your branch. Pull the latest changes and reinstall the package locally

File: src/ansys/geometry/core/designer/body.py [hunk 1]
```diff
@@ -788,6 +816,31 @@ def rotate(
             )
         )
 
+    @protect_grpc
+    @check_input_types
+    @reset_tessellation_cache
+    def get_collision(self, body: "Body") -> CollisionType:
```
Comment (RobPasMue) [inline src/ansys/geometry/core/designer/body.py:822]:
```suggestion
    def get_collision(self, body: "Body") -> CollisionType:
```

You cannot call the ``check_input_types`` header here since the class "Body" is not defined at this point. Also, you do not need to reset the tessellation with this operation since you are not modifying the underlying model.

File: src/ansys/geometry/core/designer/body.py [hunk 2]
```diff
@@ -1140,6 +1193,11 @@ def rotate(
     ) -> None:  # noqa: D102
         return self._template.rotate(axis_origin, axis_direction, angle)
 
+    @ensure_design_is_active
+    def get_collision(self, body: "Body") -> CollisionType:
+        """Get the collision state between two bodies."""
```
Comment (RobPasMue) [inline src/ansys/geometry/core/designer/body.py:1198]:
```suggestion
```

Removing since you reuse the docstring of the parent class.

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+class MeasureTools:
+    """Measure Tools for PyAnsys Geometry."""
+
+    def __init__(self, grpc_client: GrpcClient):
+        """Initialize Measure Tools class."""
+        self._grpc_client = grpc_client
+        self._measure_stub = MeasureToolsStub(self._grpc_client.channel)
+
+    def min_distance_between_objects(self, body1: "Body", body2: "Body"):
```
Comment (RobPasMue) [inline src/ansys/geometry/core/tools/measure_tools.py:67]: You have to protect this gRPC call

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    from ansys.geometry.core.designer.body import Body
+
+
+class Gap:
+    """Gap between two bodies."""
+
+    def __init__(self, grpc_client: GrpcClient, distance: float = None):
+        """Initialize Gap class."""
+        self._grpc_client = grpc_client
+        self._measure_stub = MeasureToolsStub(self._grpc_client.channel)
```
Comment (RobPasMue) [inline src/ansys/geometry/core/tools/measure_tools.py:43]: Why does the gap class have a MeasureToolsStub if you are not using it in this class? Please remove this.

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+from beartype.typing import TYPE_CHECKING
+
+from ansys.geometry.core.connection import GrpcClient
+
+if TYPE_CHECKING:  # pragma: no cover
+    from ansys.geometry.core.designer.body import Body
+
+
+class Gap:
+    """Gap between two bodies."""
```
Comment (RobPasMue) [inline src/ansys/geometry/core/tools/measure_tools.py:38]: Properly complete docstring

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+    def __init__(self, grpc_client: GrpcClient, distance: float = None):
+        """Initialize Gap class."""
+        self._grpc_client = grpc_client
+        self._measure_stub = MeasureToolsStub(self._grpc_client.channel)
+        self._distance = distance
+
+    @property
+    def distance(self):
```
Comment (RobPasMue) [inline src/ansys/geometry/core/tools/measure_tools.py:47]: Return typehint missing

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        """Get the calculated distance."""
+        return self._distance
+
+    @classmethod
+    def from_distance_response(
+        cls, grpc_client: GrpcClient, response: MinDistanceBetweenObjectsResponse
+    ):
```
Comment (RobPasMue) [inline src/ansys/geometry/core/tools/measure_tools.py:54]: Return typehint missing

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        self._measure_stub = MeasureToolsStub(self._grpc_client.channel)
+
+    def min_distance_between_objects(self, body1: "Body", body2: "Body"):
+        """Find the gap between objects."""
+        body_ids = [body1.id, body2.id]
+        response = self._measure_stub.MinDistanceBetweenObjects(
+            MinDistanceBetweenObjectsRequest(bodies=body_ids)
+        )
+        gap = Gap.from_distance_response(self._grpc_client, response)
+        return gap
```
Comment (RobPasMue) [inline src/ansys/geometry/core/tools/measure_tools.py:74]: I would get rid of the "Gap" class and just use our current ``Distance`` object.

**Review by RobPasMue — CHANGES_REQUESTED** [review]
Some comments

File: src/ansys/geometry/core/tools/measure_tools.py (hunk 8 above)
Comment (umutsoysalansys) [reply src/ansys/geometry/core/tools/measure_tools.py:74]: hey @RobPasMue The idea here is to bring other elements of the Gap method in future,  Like the vector, coordinate frame and other auxiliary methods like dx, dx, dz etc. You can see full Gap object on SpaceClaim.

Comment (RobPasMue) [reply src/ansys/geometry/core/tools/measure_tools.py:74]: Got it - but in that case, do not return floats. Return the associated PyAnsys Geometry objects: Distance, Vector3D...

``gap.distance`` is returning a float right now. It should be a ``Distance`` object

Comment (umutsoysalansys) [reply src/ansys/geometry/core/tools/measure_tools.py:74]: oh I was not aware of the distance object. Thanks.

File: src/ansys/geometry/core/tools/measure_tools.py (hunk 3 above)
Comment (umutsoysalansys) [reply src/ansys/geometry/core/tools/measure_tools.py:78]: I added the decorator @protect_grpc.

File: src/ansys/geometry/core/tools/measure_tools.py (hunk 8 above)
Comment (RobPasMue) [reply src/ansys/geometry/core/tools/measure_tools.py:85]: Yep, of course! Might be best if we set up a call for that. Propose a meeting whenever you have the chance. Ask @b-matteo to join as well in case he is interested

File: src/ansys/geometry/core/tools/measure_tools.py (hunk 3 above)
Comment (umutsoysalansys) [reply src/ansys/geometry/core/tools/measure_tools.py:78]: but I added it and now I am getting GRPC errors? 
FAILED tests/integration/test_measure_tools.py::test_min_distance_between_objects - ansys.geometry.core.errors.GeometryExitedError: Geometry service connection terminated: Sequence contains no elements

Same test is still passing locally btw

Comment (RobPasMue) [reply src/ansys/geometry/core/tools/measure_tools.py:78]: It seems to me like there IS an error on the published version at least...

FAILED tests/integration/test_measure_tools.py::test_min_distance_between_objects - ansys.geometry.core.errors.GeometryExitedError: Geometry service connection terminated: Sequence contains no elements

Looks weird. Let's try and solve that on a meeting

File: src/ansys/geometry/core/designer/body.py [hunk 9]
```diff
@@ -346,6 +358,22 @@ def rotate(
...
 
+    @abstractmethod
+    def get_collision(self, body: "Body") -> CollisionType:
+        """
+        Get the collision state between bodies.
+
+        Parameters
+        ----------
+        body: Body
+            The body object that collision state is checked with.
```
Comment (PipKat) [inline src/ansys/geometry/core/designer/body.py:369]:
```suggestion
            Object that the collision state is checked with.
```

File: src/ansys/geometry/core/designer/body.py [hunk 10]
```diff
@@ -346,6 +358,22 @@ def rotate(
...
+        Returns
+        -------
+        Collision Type
+            Enum that defines collision state between bodies.
```
Comment (PipKat) [inline src/ansys/geometry/core/designer/body.py:373]:
```suggestion
            Enum that defines the collision state between bodies.
```

File: src/ansys/geometry/core/designer/body.py [hunk 11]
```diff
@@ -788,6 +816,29 @@ def rotate(
...
 
+    @protect_grpc
+    def get_collision(self, body: "Body") -> CollisionType:
+        """
+        Get the collision state between bodies.
+
+        Parameters
+        ----------
+        body: Body
+            The body object that collision state is checked with.
```
Comment (PipKat) [inline src/ansys/geometry/core/designer/body.py:827]:
```suggestion
            Object that the collision state is checked with.
```

File: src/ansys/geometry/core/designer/body.py [hunk 12]
```diff
@@ -788,6 +816,29 @@ def rotate(
...
+        Returns
+        -------
+        Collision Type
+            Enum that defines collision state between bodies.
```
Comment (PipKat) [inline src/ansys/geometry/core/designer/body.py:831]:
```suggestion
            Enum that defines the collision state between bodies.
```

File: src/ansys/geometry/core/modeler.py [hunk 13]
```diff
@@ -399,3 +402,8 @@ def run_discovery_script_file(
     def repair_tools(self) -> RepairTools:
         """Access to repair tools."""
         return self._repair_tools
+
+    @property
+    def measure_tools(self) -> MeasureTools:
+        """Access to measure tools."""
```
Comment (PipKat) [inline src/ansys/geometry/core/modeler.py:408]:
```suggestion
        """Measurement tools."""
```

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+from ansys.geometry.core.errors import protect_grpc
+from ansys.geometry.core.misc.measurements import DEFAULT_UNITS, Distance
+
+if TYPE_CHECKING:  # pragma: no cover
+    from ansys.geometry.core.designer.body import Body
+
+
+class Gap:
+    """
+    Gap between two bodies.
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/measure_tools.py:41]:
```suggestion
    Represents a gap between two bodies.
```

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+    Parameters
+    ----------
+    distance : Distance
+        The distance between two sides of the gap.
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/measure_tools.py:46]:
```suggestion
        Distance between two sides of the gap.
```

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    @protect_grpc
+    def __init__(self, grpc_client: GrpcClient, distance: Distance = None):
+        """Initialize Gap class."""
+        self._grpc_client = grpc_client
+        self._measure_stub = MeasureToolsStub(self._grpc_client.channel)
+        self._distance = distance
+
+    @property
+    def distance(self) -> Distance:
+        """Get the calculated distance."""
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/measure_tools.py:58]: Calculated distance of what? It seems that we need a more complete description.

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+        return self._distance
+
+    @classmethod
+    @protect_grpc
+    def from_distance_response(
+        cls, grpc_client: GrpcClient, response: MinDistanceBetweenObjectsResponse
+    ) -> None:
+        """Construct Gap object from distance response."""
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/measure_tools.py:66]:
```suggestion
        """Construct a gap from the distance response."""
```
I'm not sure if the suggestion is what is meant. If we need to use Gap as a proper noun (capitalized), it should be tagged in double backticks as a code entity.

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+        distance = Distance(response.gap.distance, unit=DEFAULT_UNITS.LENGTH)
+        return cls(grpc_client, distance)
+
+
+class MeasureTools:
+    """Measure Tools for PyAnsys Geometry."""
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/measure_tools.py:72]:
```suggestion
    """Provides measurement tools for PyAnsys Geometry."""
```

File: src/ansys/geometry/core/tools/measure_tools.py [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+
+    @protect_grpc
+    def __init__(self, grpc_client: GrpcClient):
+        """Initialize Measure Tools class."""
+        self._grpc_client = grpc_client
+        self._measure_stub = MeasureToolsStub(self._grpc_client.channel)
+
+    @protect_grpc
+    def min_distance_between_objects(self, body1: "Body", body2: "Body"):
+        """Find the gap between objects."""
```
Comment (PipKat) [inline src/ansys/geometry/core/tools/measure_tools.py:82]:
```suggestion
        """Find the gap between two bodies."""
```

File: tests/integration/test_design.py [hunk 20]
```diff
@@ -57,6 +58,13 @@
...
+# TODO: re-enable when Linux service is able to use measure tools
+def skip_if_linux(modeler: Modeler):
+    """Skip test if running on Linux."""
+    if modeler.client.backend_type == BackendType.LINUX_SERVICE:
+        pytest.skip("Measure tools not available on Linux service.")
+
+
 def test_design_extrusion_and_material_assignment(modeler: Modeler):
     """Test in charge of validating the extrusion of a simple circle as a cylinder and
     assigning materials to it."""
```
Comment (PipKat) [inline tests/integration/test_design.py:70]:
```suggestion
    """Test to validate the extrusion of a simple circle as a cylinder and
    assign materials to it."""
```

File: tests/integration/test_design.py [hunk 21]
```diff
@@ -1859,3 +1867,15 @@ def test_get_active_design(modeler: Modeler):
     design1 = modeler.create_design("Design1")
     active_design = modeler.get_active_design()
     assert active_design.design_id == design1.design_id
+
+
+def test_get_collision(modeler: Modeler):
+    """Test the get_collision method of the bodies."""
```
Comment (PipKat) [inline tests/integration/test_design.py:1873]:
```suggestion
    """Test the ``get_collision()`` method of the bodies."""
```
Descriptions should generally be in plain English. Should we rewrite to: "Test to validate getting the collision of the bodies."

File: src/ansys/geometry/core/tools/measure_tools.py (hunk 18 above)
Comment (umutsoysalansys) [reply src/ansys/geometry/core/tools/measure_tools.py:72]: thanks for the reviews, how about measuring tools then? @PipKat

File: tests/integration/test_design.py (hunk 20 above)
Comment (umutsoysalansys) [reply tests/integration/test_design.py:70]: This is not my change at first place, but in spirit of better docs I am adding your suggestion here as well.

**Comment by RobPasMue** [comment]
I will check out your branch and do any final tweaks needed. Thanks @umutsoysalansys!

**Review by RobPasMue — APPROVED** [review]
LGTM! Approving and merging
